universal-login-formal-verification/src/spec.md

```diff
@@ -2,6 +2,47 @@ High-level specification of KeyHolder
 =====================================
 
 
+-   `addKey`:
+
+```act
+behaviour addKey-succ of KeyHolder
+interface addKey(address _key, uint256 _purpose)
+
+types
+
+    X : uint256
+    Y : uint256
+    Z : uint256
+
+storage
+
+        #mapping.keys[_key]  |-> _ => _purpose
+    1 + #mapping.keys[_key]  |-> X => _key |Int IrrelevantForAddressBits(X)
```

Comment:
Proposed refactorings:
`1 + #mapping.keys[_key]  |-> prevKey => asAddress(_key, prevKey)`

even better:
`1 + #mapping.keys[_key]  |-> X => asAddress(_key)`

also to consider:
`#mapping.keys[_key].key`